Lt tor or not to tor (#191)
tari-labs/tari-university: src/protocols/tor-or-not-too-tor/tor-or-not-too-tor.md

# Tor - an Overview

- [Introduction](#introduction)
  - [What is Tor?](#what-is-tor)
  - [Purpose of this Report](#purpose-of-this-report)
  - [Installation Instructions](#installation-instructions)
- [Using Tor](#using-tor)
  - [How to use Tor](#how-to-use-tor)
  - [When to use Tor](#when-to-use-tor)
  - [When not to use Tor](#when-not-to-use-tor)
- [How does Tor Network Work?](#how-does-tor-network-work)
- [Is Tor Broken?](#is-tor-broken)
- [Conclusion](#conclusion)
- [References](#references)
- [Appendices](#appendices)
  - [Appendix A: Further Investigation](#appendix-a-further-investigation)
  - [Appendix B: Links of Interest](#appendix-b-links-of-interest)
- [Contributors](#contributors)

## Introduction

### What is Tor?

Tor is a free and open source anonymity/privacy tool, meant to protect your location and identity.
The name is derived from the acronym for the original software project name, ```The Onion Router``` ([[1]], [[2]]). This 
refers to the way in which Tor protects your data, by wrapping it in multiple layers 
of encryption, similar to the layers of an onion.

> Tor uses a unique system that was developed by the US Navy to protect 

Author: Thanks! I will work this in.

> Tor uses a unique system that was developed by the US Navy to protect 

Author: Added some further clarification 1e9e7fb42008bf3f448da44a37459ada6d297584

government intelligence communications. It is part of the ```Tor Browser``` package, where the Tor Browser has been 

> configured to proxy request via the Tor network.  

Author: Nice catch, thanks.  
  
Tor is a network, that is how we route thru it. Used as a services, but it's not a protocol. Thou tools like ssh and other services can be proxied/routed via the Tor network. There a few ways to use the Tor network and I am attempting to cover the widest audience and safest suggest useage of Tor.

But I think you have at least two worthy points to try and add into my report.

> configured to proxy request via the Tor network.  

Author: Typo fixed 3a9ea61bac1550583c9a20426372d5cc7e1f78cb  
  
Other items to possible still to be addressed.

> configured to proxy request via the Tor network.  

Author: Re-word the section - 3201bb5  
  
Think it's better?

The Tor network service is the heart of the Tor project. The Tor Browser and other tools, such as OnionShare, run on top of or via the Tor network. These tools are meant to make using the Tor network as simple and secure as possible.  

### Purpose of this Report

This report gives an overview of Tor, using the Tor Browser. The author reaches a conclusion regarding whether or not to 
recommend Tor, i.e. "to Tor || !2 Tor", which means "to Tor or not to Tor" in programming language.

> ### Installation Instructions for Tor Browser

Author: Again, as I was thinking of an audience wider than just ```Tari Labs``` Engineer's, showing the ease of installing the Tor Browser and howto use, is not a waste. Plus the added point of verifying is something many people over look.

> ### Installation Instructions for Tor Browser

Reviewer: Ok cool

> ### Installation Instructions for Tor Browser

Reviewer: Cool - let's leave it in

With reference to [[1]]:
* Download the package for your operating system and language.
* Verify the package.
* Install the downloaded and verified package.

## Using Tor

> ### How to use Tor

Author: A section on ```onion``` services might be a rabbit whole on security, see if I can put something together.

> ### How to use Tor

Reviewer: Thanks - ye maybe some brief technical detail on what an onion address is - think it's OK to use the word public key  (i.e. the onion address is a hash of a randomly generated public key followed by .onion and give an example e.g duck duck go's hidden service onion address is https://3g2upl4pq6kufc4m.onion.) as described inthe format section here: here: https://www.wikiwand.com/en/.onion

> ### How to use Tor

Author: Added a new secsion on ```.onion``` address/sites - https://github.com/tari-labs/tari-university/blob/lt-tor-or-not-too-tor/src/protocols/tor-or-not-too-tor/tor-or-not-too-tor.md#what-is-an-onion-site  
  
I think this address adding more info?

> ### How to use Tor

Reviewer: Thanks

- Run the Tor Browser and visit an ```.onion``` site such as the DuckDuckGo Search engine [[3]].
- Check if you are running Tor [[4]].

### When to use Tor

The ```Tor Browser``` is an anonymity tool that blocks tracking, hides some user details and aims to make all Tor 
users look the same to average users. It could therefore be used to anonymize whistle-blowers or journalists reporting 

> on sensitive topics.

Author: Adding a notable example, would mean the anonymity was not kept?

> on sensitive topics.

Reviewer: Sorry my request was confusing - Ye, any interesting or notable real world stories which in which a tor user was deanonymized, what they did wrong etc - optional but could be interesting

> on sensitive topics.

Reviewer: See something similar here:  https://github.com/tari-labs/tari-university/commit/89b2fce1dcf5e8e82ac99b494159c22409819649 👍

### When not to use Tor

Do not use the Tor Browser if you want end-to-end encryption to a non-```.onion``` site such as your banking site, as 

> the ```Exit Node``` ends encryption and your data can be intercepted between the Exit Node and the destination server.

Author: I believe that the above statement is true, in that Tor network does not guarantee encryption once you leave the Tor network and data leaving the Tor network can be intercepted.  
  
The focus on an example site, like a banking site, that adds no benefit to using Tor, if one is expecting that Tor Browser or Tor Network will magically make your banking more secure, this is at fault.  
  
Yes, a TLS connection should not be possible too decrypt, but there are way too many other variables than just hoping Tor will secure a connection.  
  
I do thinking that I could possible state, that the Tor encryption ends at the Exit Node and point out that all other precautions for secure HTTPS or internet behaviors should still be used, like confirming that an HTTPS site has a valid SSL certificate for the site been visited. I will need to find the write words.

> the ```Exit Node``` ends encryption and your data can be intercepted between the Exit Node and the destination server.

Reviewer: Ah ok, I took it to mean that encryption was somehow terminated even when using SSL - Think those additions you mentioned will be helpful for avoiding misconceptions about Tor

> the ```Exit Node``` ends encryption and your data can be intercepted between the Exit Node and the destination server.

Author: Minor clarification 7d99637

## How does Tor Network Work?

Before Tor data enters the Tor network, it is bundled into layers of encrypted packets that are the same size. These 

> packets are then routed through a series of volunteer-operated servers called nodes, relays or bridges. Each time the 

Reviewer: Seems like the most important node type to get a very high-level understanding of is the tor exit node. Perhaps briefly mention it's job 

Something along the lines of
"A request is routed through the tor network, after a number of hops the request reached a random exit node that connects to the actual resource being requested (e.g. mywidgets.com)"

> packets are then routed through a series of volunteer-operated servers called nodes, relays or bridges. Each time the 

Author: Added some more info on nodes - f7f06d21981c8e91b57829b500abae38cf5653fb

> packets are then routed through a series of volunteer-operated servers called nodes, relays or bridges. Each time the 

Reviewer: Great thanks - I left a separate comment on one of the changes so resolving this thread

Tor data passes through one of these relays, a layer of encryption is removed to reveal the location of the next relay. 
When the data reaches the final relay on its path, known as the ```Exit Node```, the last layer of encryption is removed 
and the data is sent to its final destination.

> Each relay only decrypts enough data to learn the location of the previous and next relays. Since each path is randomly 

Author: Corrected by 3a9ea61bac1550583c9a20426372d5cc7e1f78cb  
  
Extra details possible to follow.

> generated and the relays do not keep records, it is nearly impossible for your activity to be traced back to you 

Author: This I can do, thanks.

> generated and the relays do not keep records, it is nearly impossible for your activity to be traced back to you 

Author: Two examples added - 89b2fce

through Tor’s complex network. Refer to [[5]] for more detailed information.

## Is Tor broken?

Tor is not broken if Tor services are correctly set up and if the Tor Browser is used correctly. However, it is very 
easy to do something that would make Tor appear to be broken. If you login into a remote service such as Facebook or 

> Gmail, your ```anonymity``` at this site is lost. However, the recent release of the Tor Browser package would keep 

Author: Added more info in 2bfbd45e95bf63da43906e7fcae8de94445f4db2  
  
Thou can't find anything short of a design goal, which is way too much info.

> Gmail, your ```anonymity``` at this site is lost. However, the recent release of the Tor Browser package would keep 

Reviewer: Perfect!

both sites isolated from one another.

Older Tor setups needed a user to know how to configure their proxy settings in their operating system, in order to use 
Tor services. This was very easy to get wrong or incomplete, and some user information or details could be leaked. 
Newer releases of Tor as a single package, preconfigured and shipped with a secure browser, resolved many of these 

> problems. Keeping the Tor Browser package up to date, thereby addressing any new security issues, helps to keep your 

Author: Updated by 3a9ea61bac1550583c9a20426372d5cc7e1f78cb.

> ```anonymity```.

Author: Again, more info in 2bfbd45e95bf63da43906e7fcae8de94445f4db2

> Tor has weaknesses. For example, if you are the only person using Tor on your home, office or school 

Author: Expectation of the novice users, hope I have covered it in 2bfbd45e95bf63da43906e7fcae8de94445f4db2

network, you will be discovered. This is called a time/traffic correlation attack. For an interesting talk about some 
of the Tor attacks, refer to [[6]].

## Conclusion

The following should be kept in mind:
- Anonymity is not confidentiality.
- Tor is not a Virtual Private Network (VPN).

> - Tor data leaving the Tor network can be intercepted.

Author: Are the above points always true? Yes. Data leaving the Tor Network can be captured and dumped too storage for further investigation or analysis. Any point between and including the ```Exit Node``` and the destination service could be compromised.  
  
Having weak SSL would mean that the TLS connection could be broken.  
  
Just using Tor does not guarantee anonymity nor confidentiality as stated.

> - Tor data leaving the Tor network can be intercepted.

Reviewer: Understood, think it will be good to clear up misconceptions about Tor in this report

Advantages of Tor:
- It is free and open source.
- It supports Linux, OSX and Windows.
- It is easy to install for supported operating systems.

> - Its browser is the only browser that provides secure accessibility to ```.onion``` sites.

Author: Reason why Tor Browser is the preferred tor network browser way, mostly because setting up a third-party browser to securely proxy all requests via Tor is not trivial nor are most browser focus on privacy, which the Tor Browser package is already setup and privacy focus and warns when it should be updated, a little more than the average browser.  
   
Last line is not accurate, as any browser using the Tor network should be able to access ```.onion``` sites.  
  
But, the Tor Browser is the most general of browsers, harden for privacy and generalised too improve anonymity only exceeded by using the bootable Tor distro ```Tails```  
  
Thinking maybe a link to the ChangeLog, but not sure how helpful that is?

> - Its browser is the only browser that provides secure accessibility to ```.onion``` sites.

Reviewer: Think providing a few points of the Tor Browser security/anonymity features would be interesting.

> - It is not controlled by corporation or government agencies.

Author: Thanks.

> - It is not controlled by corporation or government agencies.

Author: Corrected by 3a9ea61bac1550583c9a20426372d5cc7e1f78cb

Disadvantages of Tor:
- It is slow.

> - It does not necessarily encrypt data leaving ```Exit Node```. This data can be intercepted.

Author: This is not a specific Tor problem, but often expected or assumed, why the dire warning.

- It does not stop somebody from de-anonymizing themselves.
- It does not stop interceptors from knowing you are using Tor.
- Its network is not user-friendly, due to its secure and hidden nature.

> - Its nodes (relay/bridge) are run by volunteers.

Author: Last item updated by 3a9ea61bac1550583c9a20426372d5cc7e1f78cb  
  
Was this the only question?

on sensitive topics.

### When not to use Tor

Do not use only the Tor Browser if you expect end-to-end encryption to a non-```.onion``` site such as a banking site, as the 
```Exit Node/Relay``` ends the Tor encryption and your traffic can be intercepted between the Exit Node and the destination server.

## How does Tor Network Work?

Before Tor data enters the Tor network, it is bundled into nested layers of encrypted packets of the same size. These 
packets are then routed through a random series of volunteer-operated servers called nodes/relays. Each time the 
Tor data passes through one of these relays, a layer of encryption is removed to reveal the location of the next relay. 
When the data reaches the final relay on its path, known as the ```Exit Node/Relay```, the last layer of encryption is removed 
and the data is sent to its final destination.

There are a few different kinds of relays. A Guard/Middle (none-exit) relay, which deals with traffic inside or entering the 
Tor network. Exit relays, which are public-facing relays where Tor traffic leaves the Tor network. A special kind of relay 

> called a Bridge, which is an unlisted relay, less likely to be blocked.

Reviewer: ```suggestion
called a Bridge, which is an unlisted relay, is less likely to be blocked.
```

What is an "unlisted relay"? What does it mean to be blocked? Think a brief paragraph on listings of relay nodes would be nice - perhaps even put the node types in a table with a description of each.

Author: Thanks @sdbondi, don't want to commit you change at this time, as I think it hide your comment, which is spot on.

Each relay can only decrypt enough data to learn the location of the previous and next relay. Since each path is randomly 
generated and the relays do not keep records, it is nearly impossible for your activity to be traced back to you 
through Tor’s complex network. Refer to [[5]] for more detailed information.

More information on what a Tor Relay is or how to setup and volunteer some resources can be found at [[6]]

## What is an ```.onion``` site?

> An ```.onion``` site is a hidden service or web site that is hosted inside of the Tor network, like DuckDuckGo Search engine inside 

Reviewer: ```suggestion
An ```.onion``` address points to some resource on the Tor network called a hidden service. Hidden services are only accessible by using the Tor network. As an example,  visiting the DuckDuckGo Search engine onion address (https://3g2upl4pq6kufc4m.onion/) routes the request through the Tor network without the client knowing the host IP address of the server. The onion address is practically meaningless without it being routed through and resolved by the Tor network. 
```

Author: Thanks @sdbondi, now you will have a contributors credit too .. ;-)

the Tor network [[3]]. Generally only accessible within the Tor network. Traffic between a Tor client and an ```.onion``` site should 

> never leave the Tor network, keeping the network traffic safer than publicly hosted sites. Thou more difficult to find, as the 

Reviewer: ```suggestion
never leave the Tor network, keeping the network traffic safer than publicly hosted sites. Though more difficult to find, as the 
```

> ```.onion``` address is not an easy to remember domain name, but generally opaque, non-mnemonic, 16- or 56-character alpha-semi-numerical 

Reviewer: ```suggestion
```.onion``` address is not an easy-to-remember domain name, but generally opaque, non-mnemonic, 16- or 56-character alpha-semi-numerical 
```

public DNS server, if the ```network.proxy.socks_remote_dns``` was not set to true in FireFox. Theses DNS request could 
be used to track where a user might be surfing and thus, de-anonymize the user.

Tor is not broken if Tor services are correctly setup or if the Tor Browser is used correctly. It is very easy to do 
something that would de-anonymize a user, like using an older browser or tool that is not configured to proxy all 
traffic via Tor network services.

Another example is if the user logs into a remote service such as Facebook or Gmail, your anonymity at this site is 
lost. What many people don't know, is that other sites use tracking techniques, like cookies, could de-anonymize 
the user on other sites too. Further information about online tracking can be found at [[9]]

However, recent releases of the Tor Browser notify users of updates and also work toward keeping each site isolated 
in their own Tab or session, addressing old and possible insecure releases and user tracking via cookies.

Tor some has weaknesses, for example, if you are the only person using Tor on your home, office or school network, 

> you will be discovered.  Another is that a site knows when it's been accessed using Tor.  These shortcomings 

Reviewer: ```suggestion
you could be de-anonymized.  Another is that a site knows when it's been accessed using Tor.  These shortcomings 
```
